z_transforms

edelrosespol opened the pull request:
<code>z_transform(f,n,z)</code> and <code>inverse_z_transform(F,z,n)</code> from unilateral basic pairs and properties tables

z transform from discrete time domain function f[n] and inverse z transform from z domain using basic pairs table. z transform properties used to create other transforms found at textbook table. Concepts and exercises are presented at Signals and Systems course at [http://blog.espol.edu.ec/telg1001/transformada-z-unilateral-e-inversa-con-sympy/](http://blog.espol.edu.ec/telg1001/transformada-z-unilateral-e-inversa-con-sympy/)

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* integrals
  * New submodule <code>sympy.integrals.z_transforms</code> for unilateral z transforms.
     * Function <code>z_transform(f,n,z)</code>
     * Function <code>inverse_z_transform(F,z,n)</code>
<!-- END RELEASE NOTES -->

edelrosespol committed:
z_transforms

z_transform(f,n,z) and inverse_z_transform(F,z,n) from basic pairs and properties tables

zolabar commented:
@edelrosespol, thank you for the contribution!

I think that z-Transforms are fundamental in modern engineering, since everything today is digital, i.e. discrete.

In control/lti.py we still need z-Transfer function analysis, eventually there is room for coopertation between sympy/discrete and control?

@hanspi42, what do you think?

edelrosespol committed:
Update .mailmap

update mailmap with z_transform branch and commit.

sylee957 commented on sympy/discrete/z_transforms.py:
```diff
@@ -0,0 +1,1388 @@
+##########################################################################
+
+
+class z_transformError(NotImplementedError):
```
I would name the error class 
```suggestion
class ZTransformError(NotImplementedError):
```

edelrosespol replied:
greetings @sylee957 , thanks for your comment. 
'ZT' comes from 'ZT = z_transform' used at test files. 'z_transforms' is the name of function proposed, like the example at laplace transforms to simplify asserts. ¿name conventions to follow?
I appreciate your feedback

hanspi42 replied:
The naming convention here is that class names do not have underscores, but each word is capitalized. We follow pep8 conventions, se here: https://peps.python.org/pep-0008/#class-names

hanspi42 commented on sympy/discrete/tests/test_z_transforms.py:
```diff
@@ -0,0 +1,249 @@
+    assert ZT(-(0.5**n)*Heaviside(-n-1),n,z) == z/(z - 0.5)
+
+    # combined properties
+    assert ZT(-n*(0.5**n)*Heaviside(-n-1),n,z) == 0.5*z/(z - 0.5)**2
```
You should never use real numbers when you want rationals.  Write `S.Half` instead of `0.5`, or `S(3)/5` instead of `0.6` if it is supposed to be a rational number.

edelrosespol replied:
greetings @hanspi42
Thanks for your comments to improve the algorithm, I would like to implement something about this one.
I test ZT with S.Half and Rational(3/4) and got this:
>>> ZT((S.Half)**n*Heaviside(n),n,z)
2*z/(2*z - 1)
>>> from sympy.core.numbers import Rational
>>> ZT((Rational(3/4))**n*Heaviside(n),n,z)
4*z/(4*z - 3)
the result should be expressed as rational number.
Should be included asserts with rational numbers as a first step.
I Appreciate your comments.

hanspi42 replied:
This works by chance. If you write `3/4` in Python, it is immediately converted to `0.75`, handed over to `Rational`, and `Rational` converts it back to `3/4`. The correct way would be `Rational(3, 4)`. Try `Rational(15/123)` and `Rational(15, 123)`.

You can also do `S(3)/4`, because then Python will see a SymPy singleton object `S(3)` and will not convert anything to `0.75` when it is to be divided by `4`.

hanspi42 commented:
One more thing: you rely on `Heaviside(n)`, but this is the wrong function, as `Heaviside(0)` is `S.Half`. The z transform must use `Heaviside(n, H0=1)`.

edelrosespol committed:
Reviewed simplification for Rational poles and zeros

express F[z] with Rational Poles and Zeros. Returns full expressions when z_transforms are form simple terms.

edelrosespol committed:
trailing space removed

edelrosespol committed:
examples libraries updated

to use S.Half and S.One/2

hanspi42 commented:
It would be good to start with discussing my comments above. `discrete/transforms` or `discrete` is the wrong place for this, all transforms in the `discrete` module transform discrete signals to discrete signals. They are not integral transforms. The z Transform, however, is an integral transform, and should be in `integrals\transform`.

The reason for this is that a z function $H(z)$ is continuous, not discrete, and that it is directly connected to a Fourier function $H(j\omega)$ by the substitution $z=e^{j \omega T}$, just as a Laplace function $G(s)$ is connected to a Fourier function $G(j\omega)$ by substitution $s=j\omega$.

Before we resolve this, all other things are just cosmetic.

edelrosespol commented:
> It would be good to start with discussing my comments above. `discrete/transforms` or `discrete` is the wrong place for this, all transforms in the `discrete` module transform discrete signals to discrete signals. They are not integral transforms. The z Transform, however, is an integral transform, and should be in `integrals\transform`.

@hanspi42 got a point. Z transforms use `summation` one way to <code>z</code> and `integrals` the way back to `n`. Further development will be at `integrals` and actual proposal became simmilar to "_laplace_transform(fn, t_, s) ... It tries to apply all known rules recursively". 
In order to "move" files, the procedure should be: 'copy these files to `integrals\z_transform` and delete from these locations? keeping this branch? Let me know about procedures. Comments are appreciated.

hanspi42 commented:
You can simply move them with `git mv` on the command line or like [this](https://docs.github.com/en/repositories/working-with-files/managing-files/moving-a-file-to-a-new-location) in the web interface.

I would call the files `ztransform` and `test_ztransform`.

The next two things would be
- give it the standard structure for integral transfoms,
- review the code and functionality.

In which order would you like to do that?

sylee957 commented:
Integral transform classes also naturally support the unevaluated transforms like `LaplaceTransform(f(x), x, s)` where it can represent an algebra of unevaluated transforms.

hanspi42 commented:
> Integral transform classes also naturally support the unevaluated objects like `LaplaceTransform(f(x), x, s)` where it can represent an algebra of unevaluated transforms.

Exactly, that is the "wrapper" I'm talking about.

edelrosespol committed:
Update and rename sympy/discrete/z_transforms.py to sympy/integrals/ztransforms.py

moved to integrals for structure compatibility with further development with integrals.

edelrosespol committed:
Update and rename sympy/discrete/tests/test_z_transforms.py to sympy/integrals/test/test_ztransforms.py

moved to integrals for structure compatibility with further develoment.

hanspi42 commented:
Now the test in in the wrong sub-directory; it should be in `tests`, not `test`.

I have started looking into your code, and I immediately see two issues:

```python
from sympy.integrals.ztransforms import (z_transform, inverse_z_transform)
from sympy.abc import z, n
h1 = inverse_z_transform(1+z**(-1), z, n)
h2 = inverse_z_transform(z**(-2)*z/(z - 1), z, n)
```

Then $h_1=\delta\left(n\right) + \delta\left(n - 1\right)$, but $h_2=\theta\left(n - 2\right)$. So $h_1$ is the inverse $z$ transform as a sum of time-shifted, weighted Dirac impulses, and $h_2$ is an expression that will give the weights of the time-shifted Dirac impulses if integers are inserted for $n$.

So the first question would be: what should be the format of the time series? Dirac impulses or just numbers? From your test cases I would guess that it should not be Dirac impulses.

The second issue is the following:
```python
In [2]: print(h2.subs(n, 2))
1/2
```
It should be `1`.  This happens because `Heviside(0)` is `1/2` in SymPy, as I mentioned above.

edelrosespol committed:
Rename sympy/integrals/test/test_ztransforms.py to sympy/integrals/tests/test_ztransforms.py

moved to tests form test

edelrosespol commented:
> Now the test in in the wrong sub-directory; it should be in `tests`, not `test`.

thanks, my mistake. Moved to tests,

hanspi42 commented:
> thanks, my mistake. Moved to tests,

OK ;) I'll continue my review after you have dealt with the other two comments.

sylee957 commented:
Does this use `residue` for inverse transforms? I think that it could be good to check some applications and consistency between this and that.

hanspi42 commented:
> a sum of time-shifted, weighted Dirac impulses

Did you by any chance want this to be the following?

```
KroneckerDelta(n, 0) + KroneckerDelta(n, 1)
```

edelrosespol committed:
Review for symbolic coefficients at inverse z_transforms

Review with symbolic coefficientes and rational expressions for inverse z transform.

edelrosespol commented:
> One more thing: you rely on `Heaviside(n)`, but this is the wrong function, as `Heaviside(0)` is `S.Half`. The z transform must use `Heaviside(n, H0=1)`.

Starting to review Heaviside(0), checking with implemented Laplace_transform as a reference.

edelrosespol committed:
test_file update

test file_update

edelrosespol committed:
inverse_z_transform(F,z,n,H0=S.Half)

used to update as desired to H0=S.One and keep definition given at https://docs.sympy.org/latest/modules/functions/special.html#heaviside

Heaviside_edge_toOne(f,n) is an optional function to change expression Heaviside(n,H0=S.Half) to H0=S.One in   f(n) expression.

edelrosespol committed:
update test_ztransforms with symbol a

update assert assert ZT(Heaviside(n-a),n,z) == (z**(1-a)/(z - 1), Abs(z) > 1, True)
